crates/bevy_pbr/src/render/shadow_sampling.wgsl

```diff
@@ -126,3 +167,166 @@ fn sample_shadow_map(light_local: vec2<f32>, depth: f32, array_index: i32, texel
+    light_local: vec3<f32>,
+    depth: f32,
+    light_id: u32,
+) -> f32 {
+    return sample_shadow_cubemap_hardware(
+        light_local + position.x * x_basis + position.y * y_basis,
+        depth,
+        light_id
+    ) * coeff;
+}
+
+// This more or less does what Castano13 does, but in 3D space. Castano13 is
+// essentially an optimized 2D Gaussian filter that takes advantage of the
+// bilinear filtering hardware to reduce the number of taps needed. This trick
+// doesn't apply to cubemaps, so we manually apply a Gaussian filter over the
+// standard 8xMSAA pattern instead.
+fn sample_shadow_cubemap_gaussian(
+    light_local: vec3<f32>,
+    depth: f32,
+    texel_size: f32,
+    distance_to_light: f32,
+    light_id: u32,
+) -> f32 {
+    // Create an orthonormal basis via Gram-Schmidt normalization so we can
+    // apply a 2D sampling pattern to a cubemap.
+    var up = vec3(0.0, 1.0, 0.0);
+    if (dot(up, normalize(light_local)) > 0.99) {
+        up = vec3(1.0, 0.0, 0.0);   // Avoid creating a degenerate basis.
+    }
+
+    let scale = texel_size * distance_to_light;
+    let x_basis = normalize(cross(light_local, up)) * scale;
+    let y_basis = normalize(cross(light_local, x_basis)) * scale;
+
+    var sum: f32 = 0.0;
+    sum += tap_shadow_cubemap(
+        D3D_SAMPLE_POINT_POSITIONS[0], D3D_SAMPLE_POINT_COEFFS[0],
+        x_basis, y_basis, light_local, depth, light_id);
+    sum += tap_shadow_cubemap(
+        D3D_SAMPLE_POINT_POSITIONS[1], D3D_SAMPLE_POINT_COEFFS[1],
+        x_basis, y_basis, light_local, depth, light_id);
+    sum += tap_shadow_cubemap(
+        D3D_SAMPLE_POINT_POSITIONS[2], D3D_SAMPLE_POINT_COEFFS[2],
+        x_basis, y_basis, light_local, depth, light_id);
+    sum += tap_shadow_cubemap(
+        D3D_SAMPLE_POINT_POSITIONS[3], D3D_SAMPLE_POINT_COEFFS[3],
+        x_basis, y_basis, light_local, depth, light_id);
+    sum += tap_shadow_cubemap(
+        D3D_SAMPLE_POINT_POSITIONS[4], D3D_SAMPLE_POINT_COEFFS[4],
+        x_basis, y_basis, light_local, depth, light_id);
+    sum += tap_shadow_cubemap(
+        D3D_SAMPLE_POINT_POSITIONS[5], D3D_SAMPLE_POINT_COEFFS[5],
+        x_basis, y_basis, light_local, depth, light_id);
+    sum += tap_shadow_cubemap(
+        D3D_SAMPLE_POINT_POSITIONS[6], D3D_SAMPLE_POINT_COEFFS[6],
+        x_basis, y_basis, light_local, depth, light_id);
+    sum += tap_shadow_cubemap(
+        D3D_SAMPLE_POINT_POSITIONS[7], D3D_SAMPLE_POINT_COEFFS[7],
+        x_basis, y_basis, light_local, depth, light_id);
+    return sum;
+}
+
+// This is a port of the Jimenez14 filter above to the 3D space. It jitters the
+// points in the spiral pattern after first creating a 2D orthonormal basis
+// along the principal light direction.
+fn sample_shadow_cubemap_temporal(
+    light_local: vec3<f32>,
+    depth: f32,
+    texel_size: f32,
+    distance_to_light: f32,
+    light_id: u32,
+) -> f32 {
+    // Create an orthonormal basis via Gram-Schmidt normalization so we can
```

Comment:
Maybe put this in a reusable function given a normal and up vector.

Reply:
It's been moved to a new function `orthonormalize`.